### PR TITLE
EPMUII-7058 equipment categories page layout

### PR DESCRIPTION
**Pull request by ALLaTk**
Task: EPMUII-7058 Equipment categories - catalog landing page - Page layout
  - add the equipment categories how a landing page of the catalog 
  - add the equipment categories store
  
Screenshot: 
categories

### CODE REVIEW

File: src/app/catalog/catalog-routing.module.ts [hunk 1]
```diff
@@ -1,12 +1,12 @@
 import { NgModule } from '@angular/core';
 import { RouterModule, Routes } from '@angular/router';
-import { CatalogComponent } from './containers';
 import { EquipmentItemComponent } from '@app/catalog/containers';
+import { CategoriesComponent } from './containers/categories/categories.component';
 
 const routes: Routes = [
   {
     path: '',
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/catalog-routing.module.ts:8]: Done. But I'm not sure I got it right

File: src/app/catalog/containers/category-item/category-item.component.html [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<section class="category-item" >
+  <div #image class="category-item__img"></div>
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/category-item/category-item.component.html]: Done

File: src/app/catalog/models/categories.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export interface Category {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/models/categories.ts]: Done

File: src/app/catalog/services/api/api.service.ts [hunk 4]
```diff
@@ -27,4 +28,9 @@ export class ApiService {
       responseType: 'arraybuffer',
     });
   }
+
+  getCategories(): Observable<BaseItemsResponse<Category>> {
+    const params = new HttpParams().set('limit', 1000);
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/services/api/api.service.ts:34]: Done

File: src/app/catalog/services/controller/controller.service.ts [hunk 5]
```diff
@@ -32,4 +37,22 @@ export class ControllerService {
...
   }
+
+  public getCategories() {
+    this.api
+      .getCategories()
+      .pipe(tap((res) => this.store.dispatch(new GetCategories(res.items))))
+      .subscribe();
+  }
+
+  // getCategories() {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/services/controller/controller.service.ts]: Done

File: src/app/catalog/services/controller/controller.service.ts [hunk 6]
```diff
@@ -32,4 +37,22 @@ export class ControllerService {
   getPhotoById(photoId: string): Observable<Blob> {
     return this.api.getPhotoById(photoId).pipe(map((res) => new Blob([res], { type: 'image/jpeg' })));
   }
+
+  public getCategories() {
+    this.api
+      .getCategories()
+      .pipe(tap((res) => this.store.dispatch(new GetCategories(res.items))))
+      .subscribe();
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/services/controller/controller.service.ts]: Done

File: src/app/catalog/components/category-item/category-item.component.scss [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+:host {
+  flex: 1 1 auto;
+}
+
+.category-item {
+  background: $color-bg-base;
+  border-radius: 30px;
+  height: 100%;
+  &-img {
```
Comment (vyMoon) [inline src/app/catalog/components/category-item/category-item.component.scss:11]: add empty line before this

File: src/app/catalog/components/category-item/category-item.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { CategoryItemComponent } from './category-item.component';
+
+describe('CategoryItemComponent', () => {
```
Comment (vyMoon) [inline src/app/catalog/components/category-item/category-item.component.spec.ts:5]: add x to have xdesribe here

File: src/app/catalog/components/category-tab/category-tab.component.html [hunk 9]
```diff
@@ -0,0 +1,2 @@
+<button class="category-equipment-button" routerLink="{{ '/catalog' | categoryPath: item?.id }}"
```
Comment (vyMoon) [inline src/app/catalog/components/category-tab/category-tab.component.html:1]: let's change the tag to a. Since it is a navigation stuf.

Comment (vyMoon) [reply src/app/catalog/components/category-tab/category-tab.component.html:1]: I think the pipe should take item and return string and '/catalog' should be defined in the pipe.

File: src/app/catalog/components/category-item/category-item.component.scss (hunk 7 above)
Comment (ALLaTk) [reply src/app/catalog/components/category-item/category-item.component.scss:12]: done

File: src/app/catalog/components/category-item/category-item.component.scss [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    border-radius: 10px;
+    display: block;
+    margin: 0;
+    height: 165px;
+    width: 100%;
+    background: #F7F7F7;
+    cursor: pointer;
+  }
+  &-title {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/components/category-item/category-item.component.scss:22]: done

File: src/app/catalog/components/category-item/category-item.component.spec.ts (hunk 8 above)
Comment (ALLaTk) [reply src/app/catalog/components/category-item/category-item.component.spec.ts]: done

File: src/app/catalog/components/category-tab/category-tab.component.html (hunk 9 above)
Comment (ALLaTk) [reply src/app/catalog/components/category-tab/category-tab.component.html]: done

File: src/app/catalog/containers/category-set/category-set.component.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+  ngAfterViewInit() {
+    if (!this.categoriesElements) return;
+
+    const activeCategory: ElementRef<HTMLDivElement> = this.categoriesElements['_results'].find(
+      (el: ElementRef<HTMLDivElement>) => {
+        return el.nativeElement.classList.contains('active');
+      },
+    );
+    const activeCategoryX: number = activeCategory.nativeElement.getBoundingClientRect().x;
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/category-set/category-set.component.ts]: done

File: src/app/catalog/pipes/category-path/category-path.pipe.spec.ts [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import { CategoryPathPipe } from './category-path.pipe';
+
+describe('CategoryPathPipe', () => {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/pipes/category-path/category-path.pipe.spec.ts]: done

File: src/app/catalog/resolver/categories.resolver.spec.ts [hunk 13]
```diff
@@ -0,0 +1,16 @@
+import { TestBed } from '@angular/core/testing';
+
+import { CategoriesResolver } from './categories.resolver';
+
+describe('CategoriesResolver', () => {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/resolver/categories.resolver.spec.ts]: done

File: src/app/catalog/services/controller/controller.service.ts [hunk 14]
```diff
@@ -8,20 +9,21 @@ import { CatalogState, GetCatalog } from '../../store';
 @Injectable()
 export class ControllerService {
   @Select(CatalogState.catalog) catalog$!: Observable<Equipment[]>;
+  apiSearch: any;
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/services/controller/controller.service.ts]: deleted as it is not used anywhere else

File: src/app/catalog/components/category-tab/category-tab.component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,24 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { CategoryTabComponent } from './category-tab.component';
+
+describe('CategoryTabComponent', () => {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/components/category-tab/category-tab.component.spec.ts]: done

File: src/app/catalog/containers/categories/categories.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { CategoriesComponent } from './categories.component';
+
+describe('CategoriesComponent', () => {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/categories/categories.component.spec.ts]: done

File: src/app/catalog/containers/category-set/category-set.component.scss [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+  overflow-x: hidden;
+}
+
+.category-equipment {
+  display: flex;
+  height: 100%;
+  padding: 25px 25px 25px 25px;
+  overflow-x: scroll;
+  scroll-snap-type: x mandatory;
+  &-list {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/category-set/category-set.component.scss:17]: done

File: src/app/catalog/containers/category-set/category-set.component.scss [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    display: flex;
+    gap: 10px;
+  }
+  &-item {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/category-set/category-set.component.scss:22]: done

File: src/app/catalog/containers/category-set/category-set.component.spec.ts [hunk 19]
```diff
@@ -0,0 +1,24 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { CategorySetComponent } from './category-set.component';
+
+describe('CategorySetComponent', () => {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/category-set/category-set.component.spec.ts]: done

File: src/app/catalog/containers/catalog/catalog.component.ts [hunk 20]
```diff
@@ -10,17 +11,25 @@ import { MainPageHeaderService } from '@shared/services/main-page-header.service
...
+  ngOnInit() {
+    this.route.params.subscribe((param) => {
+      param['categoryId']
+        ? this.controller.filterEquipmentByCategory(Number(param['categoryId']))
+        : this.controller.getCatalog();
+    });
   }
 
-  public onSearch(term: string) {
+  onSearch(term: string) {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/catalog/catalog.component.ts:32]: It's not mine

File: src/app/catalog/containers/categories/categories.component.ts [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+  selector: 'lc-categories',
+  templateUrl: './categories.component.html',
+  styleUrls: ['./categories.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class CategoriesComponent implements OnInit {
+  categories: Category[] = [];
+
+  constructor(private mainPageHeaderService: MainPageHeaderService, private route: ActivatedRoute) {
+    mainPageHeaderService.setPageTitle('Категории оборудования');
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/categories/categories.component.ts:16]: done

File: src/app/catalog/containers/category-set/category-set.component.ts [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+      has_subcategory: false,
+      id: 0,
+      max_reservation_time: 0,
+      max_reservation_units: 0,
+      name: 'Все',
+    },
+  ];
+  selectedCategoryId: number = Number(this.route.snapshot.params['categoryId']);
+
+  private marginLeft = 25;
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/category-set/category-set.component.ts:35]: done

File: src/app/catalog/containers/category-set/category-set.component.ts [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+  @ViewChildren('categoryElement') categoriesElements?: QueryList<ElementRef<HTMLDivElement>>;
+
+  categories: Category[] = [
+    {
+      has_subcategory: false,
+      id: 0,
+      max_reservation_time: 0,
+      max_reservation_units: 0,
+      name: 'Все',
+    },
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/category-set/category-set.component.ts:31]: done

[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/category-set/category-set.component.ts:31]: done

File: src/app/catalog/containers/catalog/catalog.component.ts [hunk 24]
```diff
@@ -10,17 +11,25 @@ import { MainPageHeaderService } from '@shared/services/main-page-header.service
...
     mainPageHeaderService.setPageTitle('Каталог');
   }
 
-  ngOnInit(): void {
-    this.controller.getCatalog();
+  ngOnInit() {
+    this.route.params.subscribe((param) => {
+      param['categoryId']
+        ? this.controller.filterEquipmentByCategory(Number(param['categoryId']))
+        : this.controller.getCatalog();
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/catalog/catalog.component.ts:28]: done

File: src/app/catalog/containers/category-set/category-set.component.ts [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+  private marginLeft = 25;
+
+  constructor(private route: ActivatedRoute) {}
+
+  ngOnInit() {
+    if (!this.selectedCategoryId) {
+      this.selectedCategoryId = 0;
+    }
+
+    this.categories = this.categories.concat(this.route.snapshot.data['activeCategories']);
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/category-set/category-set.component.ts:44]: done

File: src/app/catalog/containers/catalog/catalog.component.ts [hunk 26]
```diff
@@ -10,17 +11,25 @@ import { MainPageHeaderService } from '@shared/services/main-page-header.service
...
+    private mainPageHeaderService: MainPageHeaderService,
+    private route: ActivatedRoute,
+  ) {
     mainPageHeaderService.setPageTitle('Каталог');
   }
 
-  ngOnInit(): void {
-    this.controller.getCatalog();
+  ngOnInit() {
+    this.route.params.subscribe((param) => {
```
[the remark this answers is not in the document]
Comment (ALLaTk) [reply src/app/catalog/containers/catalog/catalog.component.ts:25]: done

**Review by vyMoon — APPROVED** [review]

**Pull request merged**